docs/migration-v2/index.md

```diff
@@ -0,0 +1,237 @@
+  hooks: {
+    "&:hover": "&:hover",
+  },
+  fallback: "revert-layer",
+  debug: true,
+});
+```
+
+### Updating the `sort` option
+
+If you previously used `sort: true`, you can simply remove that option, as it is
+now enabled by default.
+
+Otherwise, you can add the following to your configuration to disable sorting,
+which most closely resembles the default behavior in v1:
+
+```typescript
+sort: {
+  properties: false,
+  conditionalStyles: false
+}
+```
+
+For more information about sorting, please review the configuration guide.
+
+## Setup
+
+In v1, the `createHooks` function returned a tuple containing (1) a CSS string
+(the style sheet needed to support the configured hooks) and (2) the `css`
+function used to create inline style rules. In v2, it returns an object instead.
+This can be destructured in a similar manner.
+
+### Destructuring the `createHooks` return value
+
+Change the array (tuple) destructuring syntax to use object destructuring syntax
+instead. Note that what was previously called `hooks` (by convention) has been
+renamed to `styleSheet`.
+
+#### Before
+
+```typescript
+// src/css.ts
+
+export const [hooks, css] = createHooks(/* ... */);
+```
+
+#### After
+
+```typescript
+// src/css.ts
+
+export const { styleSheet, css } = createHooks(/* ... */);
+```
+
+### Adding the style sheet
+
+Find where the style sheet is rendered, e.g. in your root component. Update the
+`hooks` import to the new name `styleSheet`, and invoke it as a function to
+render the CSS string.
+
+#### Before
+
+```tsx
+// src/app.tsx
+
+import { hooks } from "./css";
+
+export function App() {
+  return (
+    <>
+      <style dangerouslySetInnerHTML={{ __html: hooks }} />
+      <HomePage />
+    </>
+  );
+}
+```
+
+#### After
+
+```tsx
+// src/app.tsx
+
+import { styleSheet } from "./css";
+
+export function App() {
+  return (
+    <>
+      <style dangerouslySetInnerHTML={{ __html: styleSheet() }} />
+      <HomePage />
+    </>
+  );
+}
+```
+
+### Usage
+
+The most prominent change in v2 is a more advanced syntax for conditional
+styles, which allows hooks to be combined and reused more effectively. In v1,
+nesting provided the means for hook composition, but this was strictly an "and"
+operation. In v2, a `match` callback replaces nested style objects to enable
+"or" and "not" operations—a slightly heavier, but much more powerful, syntax.
+
+Wherever you use the `css` function, you'll need to migrate to the `match`
+callback.
+
+#### Basic use case
+
+##### Before
+
+```jsx
+export function Button({ children }) {
+  return (
+    <button
+      style={css({
+        color: "blue",
+        "&:hover": {
+          color: "red",
+        },
+      })}
+    >
+      {children}
+    </button>
+  );
+}
+```
+
+##### After
+
+```jsx
+export function Button({ children }) {
+  return (
+    <button
+      style={css({
+        color: "blue",
+        match: on => [
+          on("&:hover", {
+            color: "red",
+          }),
+        ],
```

Comment:
My only concern here is the syntax complexity. The "before" is simple and easy to grasp.

I do wonder if it's possible to still allow non-nested matches as currently:

```js
css({
  "&:hover": { color: "red" },
})
```

With an `all` helper dynamic prop?

```js
css({
  [all("&:enabled", "&:hover")]: { color: "blue" }
})
```

Is there a doc for how `match` works in greater depth?

Reply:
Thanks for having an eye on this and sharing your thoughts @atomiks. I'm in a bit of a rush right at the moment, so - apologies - this explanation will be brief (feel free to follow up with questions!). For conditional styles, we basically need a list of tuples (or "entries", if you want to think of them like a record). The first item being the condition; the second being the style object that applies when that condition is true. The [`Condition` type](https://github.com/css-hooks/css-hooks/blob/v2/packages/core/src/index.d.ts#L15-L19) represents the structure of the condition - think of the `S` arg as a union of all of the configured hook names. I'm very interested in finding the simplest way to build that structure, so please keep me posted if any ideas come to mind. :-)